### PR TITLE
feat: add gnss/imu localizer 

### PR DESCRIPTION
**Pull request by rsasaki0109**
# Description
The following PR for autoware.universe changes for eagleye (GNSS/IMU) had a problem with the commit log, so I created another PR.
https://github.com/autowarefoundation/autoware.universe/pull/2848

This is the PR for integrating GNSS/IMU localizar discussed in the discussion below.
https://github.com/orgs/autowarefoundation/discussions/3257

- ~Added gnss imu localizar related launch,param~
- ~Added `pose_estimator_mode`(lidar, gnss),`twist_estimator_mode`(gyro_odom_fusion, gyro_odom_gnss_fusion) parameter in `localization.launch.xml` to switch between pose,twist~
- ~In gyro_odom_gnss_fusion mode, twist is corrected by gnss.~
- ~You can adjust the eagleye pose output in `launch/tier4_localization_launch/launch/eagleye/fix2pose.xml`.~

The above conditional branching for each pose_estimator and twist_estimator within the `tier4_localization_launch` package is complex and has been changed.
Instead, we have introduced the `map4_localization_launch` packages.
A discussion on this is provided below.
(This is a private social networking site for the TIEV IV group.)
https://map4-inc.slack.com/archives/CEL0DR2S2/p1680144103464929?thread_ts=1680143206.632879&cid=CEL0DR2S2

Switch localization launch for each structure in `map4_localization_component.launch.xml`.

localization launch  | twist estimator | pose estimator
-- | -- | --
tier4_localization_launch| gyro_odometry | ndt_scan_matcher
map4_localization_launch/eagleye_twist_localization_launch| eagleye_rt(gyro/odom/gnss fusion) | ndt_scan_matcher
map4_localization_launch/eagleye_pose_twist_localization_launch| eagleye_rt(gyro/odom/gnss fusion) | eagleye_rt(gyro/odom/gnss fusion) 

<br class="Apple-interchange-newline">

# Related links

https://github.com/autowarefoundation/autoware/pull/3261
https://github.com/autowarefoundation/autoware_launch/pull/200
https://github.com/autowarefoundation/autoware-documentation/pull/334

# Tests performed
Localizaiton methods are expected to work as normal, even if they change.
[Sample rosbag](https://drive.google.com/file/d/1VnwJx9tI3kI_cTLzP61ktuAJ1ChgygpG/view) is available for simple operation tests.
Also, if you have a rosbag of a system with the same configuration of sensors, you can check the operation to some extent.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

-  [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The Reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by rsasaki0109** [comment]
Mention the previous reviewer who is not in the "Reviewers".
@yukkysaito @KYabuuchi @kminoda

File: launch/tier4_localization_launch/launch/util/util.launch.xml
Comment (KYabuuchi) [inline launch/tier4_localization_launch/launch/util/util.launch.xml:8]: `ndt_enabled` is defined but not used. 
Currently, `ndt_enabled` is set by [pose_initializer.param.yaml](https://github.com/autowarefoundation/autoware_launch/blob/34e25d982e708fc73d454f531237b156f5f2126e/autoware_launch/config/localization/pose_initializer.param.yaml)

File: launch/tier4_localization_launch/launch/eagleye/fix2pose.xml
Comment (kminoda) [inline launch/tier4_localization_launch/launch/eagleye/fix2pose.xml:1]: Please make the file name's suffix ".launch.xml", unless there is a strong reason to avoid doing so.

File: launch/tier4_localization_launch/launch/twist_estimator/twist_estimator.launch.xml
Comment (kminoda) [inline launch/tier4_localization_launch/launch/twist_estimator/twist_estimator.launch.xml:11]: As this conditional branch may appear a bit confusing, can you describe somewhere to clarify this design?
- e.g.) Add a comment here
- e.g.) Add something like a table ([maybe like this](https://github.com/autowarefoundation/autoware.universe/pull/2848#issuecomment-1461310294)) in readme.

**Review by kminoda — COMMENTED** [review]
I left some minor comments :pray:

File: launch/tier4_localization_launch/launch/twist_estimator/twist_estimator.launch.xml
Comment (rsasaki0109) [reply launch/tier4_localization_launch/launch/twist_estimator/twist_estimator.launch.xml:1]: Fixed

File: launch/tier4_localization_launch/launch/util/util.launch.xml
Comment (rsasaki0109) [reply launch/tier4_localization_launch/launch/util/util.launch.xml]: I had overlooked that change. Fixed.

File: launch/tier4_localization_launch/launch/eagleye/fix2pose.xml
Comment (rsasaki0109) [reply launch/tier4_localization_launch/launch/eagleye/fix2pose.xml]: Fixed.

File: launch/map4_localization_launch1/CMakeLists.txt
Comment (rsasaki0109) [reply launch/map4_localization_launch1/CMakeLists.txt:2]: > Can you briefly explain why you named map4_localization_launch1, and not something like map4_localization_launch?

I wanted to clarify that there are multiple map4 localization launches, but I don't have a strong argument.

> It would be nice if you can summarize why you added map4_localization_launch instead of sticking to tier4_localization_launch. (Maybe also adding an internal link to the slack discussion thread)

OK
I am going to summarize below.
https://github.com/autowarefoundation/autoware.universe/pull/3063#issue-1620972228

Comment (kminoda) [reply launch/map4_localization_launch1/CMakeLists.txt:2]: Cool, thanks!

How about just creating `map4_localization_launch` and write all the if-conditions (for eagleye_pose_twist and eagleye_twist configuration) there?

Comment (rsasaki0109) [reply launch/map4_localization_launch1/CMakeLists.txt:2]: @kminoda 
Just to confirm, is my understanding correct that that is proposal X?

proposal X
 - map4_localization_launch package
   - eaglleye_twist_localization_launch directry
     - localization.launch.xml
   - eagleye_pose_twist_localiation_launch directry
     - localization.launch.xml
 
proposal Y
 - map4_localization_launch
   - localization.launch.xml (Conditional branching with parameter `pose_estimator_mode`.<-This is complicated.)

Comment (kminoda) [reply launch/map4_localization_launch1/CMakeLists.txt:2]: My intention was Plan Y, but either is fine for me (since it's "map4_localization_launch")

Comment (rsasaki0109) [reply launch/map4_localization_launch1/CMakeLists.txt]: I fixed it.

File: launch/tier4_localization_launch/launch/twist_estimator/twist_estimator.launch.xml
Comment (rsasaki0109) [reply launch/tier4_localization_launch/launch/twist_estimator/twist_estimator.launch.xml]: Added description to README.md of map4_localization_launch.

**Comment by KYabuuchi** [comment]
@rsasaki0109 Thanks for reflecting our review over a long term. But I have an additional request. :pray: 
There are some duplicate launch files in `map4_localization_launch`, can you change it to use the launch files in `tier4_localization_launch` instead of these?
* localization_error_monitor
* pose_estimator
* pose_twist_fusion_filter
* util

I think it is hard to maintain multiple launch files every time these packages are changed.

**Comment by rsasaki0109** [comment]
@KYabuuchi 
Your point is well taken. I have made the necessary corrections

**Comment by KYabuuchi** [comment]
Thank you for reflecting my review!  :+1: 
I think this PR has become simple enough.
In my opinion, this PR will be ready when the associated PRs (documentation and autoware_launch) reflect the current launch file structure.

**Comment by rsasaki0109** [comment]
@KYabuuchi I have already reflected it in autoware_launch. Please wait a little for the modification in autoware-documentation.

**Comment by rsasaki0109** [comment]
documentation was also updated

File: launch/map4_localization_launch/README.md
Comment (kminoda) [inline launch/map4_localization_launch/README.md:9]: It is kind of confusing here where it mentions "two" while in the next table there are three rows. How about this?
```suggestion
In addition to the LiDAR-based solution provided by `tier4_localization_launch` (shown in the first row of the following table), there are two ways to utilize eagleye in `map4_localization_launch` with the autoware localization stack.
```

**Review by kminoda — COMMENTED** [review]
Mostly LGTM, but left some minor comments.

Also, please make sure that you've got an agreement from architects about this specific directory structure, as IMHO this is the most challenging (and also the most interesting) part where we have to design carefully so that the contributers can easily add their implementation.

File: launch/map4_localization_launch/eagleye_pose_twist_localization_launch/localization_launch.drawio.svg
Comment (rsasaki0109) [reply launch/map4_localization_launch/eagleye_pose_twist_localization_launch/localization_launch.drawio.svg:1]: Fixed

File: launch/map4_localization_launch/README.md
Comment (rsasaki0109) [reply launch/map4_localization_launch/README.md]: readme.md has been corrected.

Comment (rsasaki0109) [reply launch/map4_localization_launch/README.md]: @yukkysaito @mitsudome-r 
Based on the discussion in the following link, changes have been made to the launch structure. Could you please check if there are any issues and provide your opinion?
https://github.com/autowarefoundation/autoware.universe/pull/3063#issue-1620972228

File: launch/map4_localization_launch/eagleye_pose_twist_localization_launch/localization_launch.drawio.svg
Comment (rsasaki0109) [reply launch/map4_localization_launch/eagleye_pose_twist_localization_launch/localization_launch.drawio.svg:1]: Fixed

File: launch/map4_localization_launch/README.md
Comment (rsasaki0109) [reply launch/map4_localization_launch/README.md]: @yukkysaito @mitsudome-r 
How about this PR?

Comment (yukkysaito) [reply launch/map4_localization_launch/README.md]: It's ok :+1: 

@mitsudome-r How about you?

File: localization/pose_initializer/launch/pose_initializer.launch.xml
Comment (mitsudome-r) [inline localization/pose_initializer/launch/pose_initializer.launch.xml:4]: Maybe it's better to follow the naming convention discussed [here](https://github.com/orgs/autowarefoundation/discussions/3171#discussioncomment-4597052).

Also, are there any reason why you don't specify a default value for the argument?

```suggestion
  <arg name="sub_gnss_pose_cov"/>
```

File: launch/map4_localization_launch/README.md
Comment (rsasaki0109) [reply launch/map4_localization_launch/README.md]: Fixed

File: localization/pose_initializer/launch/pose_initializer.launch.xml
Comment (rsasaki0109) [reply localization/pose_initializer/launch/pose_initializer.launch.xml]: @mitsudome-r 
No reason. Default value added.

File: launch/map4_localization_launch/README.md
Comment (rsasaki0109) [reply launch/map4_localization_launch/README.md]: @mitsudome-r I am just confirming, ok?

**Review by mitsudome-r — APPROVED** [review]
LGTM, but we should also have approval from Localization engineers before merging.

File: launch/map4_localization_launch/README.md
Comment (rsasaki0109) [reply launch/map4_localization_launch/README.md]: @kminoda 
Hello. As per the comments in this PR, I have just received an approval from @yukkysaito @mitsudome-r .

**Review by yukkysaito — APPROVED** [review]
LGTM

File: launch/map4_localization_launch/package.xml
Comment (YamatoAndo) [inline launch/map4_localization_launch/package.xml:26]: Please list the packages that depend on map4_localization_launch correctly.

Comment (rsasaki0109) [reply launch/map4_localization_launch/package.xml:29]: @YamatoAndo Fixed.

**Review by YamatoAndo — APPROVED** [review]

**Review by KYabuuchi — APPROVED** [review]
I tested all localizers(NDT, NDT+eagleye, eagleye) again, and they worked well.

**Pull request merged**
